src/bootstrap/src/core/build_steps/test.rs

```diff
@@ -355,50 +358,394 @@ impl Step for RustAnalyzer {
     const DEFAULT: bool = true;
 
     fn should_run(run: ShouldRun<'_>) -> ShouldRun<'_> {
-        run.path("src/tools/rust-analyzer")
+        run.path("src/tools/rust-analyzer").alias("rust-analyzer")
     }
 
     fn make_run(run: RunConfig<'_>) {
         run.builder.ensure(Self { stage: run.builder.top_stage, host: run.target });
     }
 
-    /// Runs `cargo test` for rust-analyzer
+    #[cfg_attr(
+        feature = "tracing",
+        instrument(
+            level = "debug",
+            name = "RustAnalyzer::run",
+            skip_all,
+            fields(stage = self.stage, host = ?self.host),
+        ),
+    )]
+    fn run(self, builder: &Builder<'_>) {
+        debug!(stage = self.stage, host = ?self.host, "ensuring compiler");
+        let compiler = builder.compiler(self.stage, self.host);
+
+        debug!(stage = self.stage, host = ?self.host, "ensuring std");
+        builder.ensure(compile::Rustc::new(compiler, self.host));
+
+        let mut cargo = tool::prepare_tool_cargo(
+            builder,
+            compiler,
+            Mode::ToolRustc,
+            self.host,
+            Kind::Test,
+            "src/tools/rust-analyzer",
+            SourceType::InTree,
+            &["in-rust-tree".to_owned()],
+        );
+        cargo.allow_features(tool::RustAnalyzer::ALLOW_FEATURES);
+
+        // RA's test suite tries to write to the source directory, that can't work in Rust CI.
+        cargo.env("SKIP_SLOW_TESTS", "1");
+
+        // NOTE: unlike `proc-macro-srv` step, we must **not** set `CARGO_WORKSPACE_DIR` because I
+        // believe `src/tools/rust-analyzer/.cargo/config.toml`'s relative `CARGO_WORKSPACE_DIR`
+        // takes effect,
+
+        cargo.add_rustc_lib_path(builder);
+
+        // NOTE: we need to skip `src/tools/rust-analyzer/xtask` as they seem to exercise rustup /
+        // stable rustfmt.
+        //
+        // NOTE: you can only skip a specific workspace package via `--skip=...` if you *also*
+        // specify `--workspace`.
+        cargo.arg("--workspace");
+        cargo.arg("--exclude=xtask");
+
+        let mut skip_tests = vec![
+            // FIXME: may need a fix from https://github.com/rust-lang/rust-analyzer/pull/19124.
+            "config::tests::cargo_target_dir_subdir",
+            // FIXME: I think this exercises cargo somehow?
+            "tests::smoke_test_real_sysroot_cargo",
```

Comment:
This invokes `rustc --print sysroot` and then runs `cargo metadata` on the sysroot

Reply:
It's possible that `rustc` doesn't go through the rustc wrapper, and maybe invoking a wrong rustc or sth, which maybe reports a wrong/different sysroot?

Reply:
Yes, from a quick glance we seem to consider the following for rustc (in that order):
- Check for the `RUSTC` env var
- Check if `rustc` is in `PATH`
- Check if `rustc` exists as a proxy in `<cargo-home>/bin/rustc`

Reply:
I think what I'll end up doing is to open a tracking issue for `./x test rust-analyzer` (in rust-lang/rust), and try to land a minimal version (with most of these skips) to make it possible to run try-jobs against r-l/r's CI, and track some of these tests being disabled separately.

Reply:
Yes, I don't think we want to fix all of these tests. Most of these are const eval related and that infra is semi broken anyways (I imagine we make 64 bit layout assumptions in some places accidentally)

Reply:
Yeah makes sense. I'll wait for #137080 to land first, then I'll pull the actual r-a test bits out into a separate PR.